### PR TITLE
Added history like object as in Keras

### LINKED ISSUE #3: Add history like object as in Keras for plotting
## 🚀 Feature

Currently, we return metrics in the train_step and val_step.
Maybe we can use them to create a dictionary of a dictionary containing a list

```history = {"train" : {"top1_acc" : [] , "top5_acc" : [], "loss" : [] }, "val" : {"top1_acc" : [] , "top5_acc" : [], "loss" : [] }}```

Plus we can provide plotting function for this history using matplotlib too.

### PR DESCRIPTION
**Pull request by premalrupnur**
Closes #3

### CODE REVIEW

File: examples/train_step_example.py [hunk 1]
```diff
@@ -52,19 +52,13 @@
     criterion = (nn.CrossEntropyLoss())  # All classification problems we need Cross entropy loss
     early_stopper = utils.EarlyStopping(patience=7, verbose=True, path=SAVE_PATH)
```
Comment (oke-aditya) [inline examples/train_step_example.py]: You should probably not even touch this file.
This is example which was created to show `train_step`. This is going to be part of documentation and kills the purpose of file.
Please use a seperate `dev.py` file and do not commit it / push to github.

File: pytorch_cnn_trainer/engine.py [hunk 2]
```diff
@@ -72,6 +72,9 @@ def train_step(
     top1_m = utils.AverageMeter()
     top5_m = utils.AverageMeter()
     cnt = 0
+    losses_t = []
```
Comment (oke-aditya) [inline pytorch_cnn_trainer/engine.py]: May I know how this solves the problem?
Why would you collect every `train_step` metrics?
Train_step was returning them anyways can we use that directly to `fit()` ?

File: pytorch_cnn_trainer/engine.py [hunk 3]
```diff
@@ -188,6 +194,9 @@ def val_step(
     top1_m = utils.AverageMeter()
     top5_m = utils.AverageMeter()
     cnt = 0
+    losses_v = []
```
Comment (oke-aditya) [inline pytorch_cnn_trainer/engine.py]: Again same comment from `train_step`. I'm already returning metrics on every `step` of training.
I want metrics for every `epoch` how does logging here solve the issue?

File: pytorch_cnn_trainer/engine.py [hunk 4]
```diff
@@ -347,8 +362,8 @@ def fit(
                 break
 
         print("Done Training, Model Saved to Disk")
-
-    return True  # For now, we should probably return an history object like keras.
+    history = {"train":train_metrics, "val" : valid_metrics}
```
Comment (oke-aditya) [inline pytorch_cnn_trainer/engine.py]: What `train_metrics` are coming here?
If you see already metrics per epoch were being returned. It would be wiser to edit only `fit()` function to get metrics logged.

I think you misunderstood the idea.

Idea was only edit `fit()` which already has metrics, and they could be simply logged in history dictionary.

**Review by oke-aditya — CHANGES_REQUESTED** [review]
Please think about points I have pointed.
Let me know your thoughts as well.

File: examples/train_step_example.py (hunk 1 above)
Comment (premalrupnur) [reply examples/train_step_example.py]: ohh sorry actually I did it for testing purpose  but anyways why to use 2 different functions if engine.fit can be used for train and valid set

File: pytorch_cnn_trainer/engine.py (hunk 2 above)
Comment (premalrupnur) [reply pytorch_cnn_trainer/engine.py]: actually I did'nt try printing train_metrics  so I thought it was not storing values of metrics but just updating a variable and value for previous epoch would be lost so I used list to store each

File: examples/train_step_example.py (hunk 1 above)
Comment (oke-aditya) [reply examples/train_step_example.py]: There is deeper reason why `train_step` and `val_step` are present.
The reason being, it gives really powerful control and allows a lot of in between operations. Have a look at examples.

File: pytorch_cnn_trainer/engine.py (hunk 2 above)
Comment (oke-aditya) [reply pytorch_cnn_trainer/engine.py]: It actually stores metrics in oredered dict.

File: pytorch_cnn_trainer/engine.py (hunk 4 above)
Comment (premalrupnur) [reply pytorch_cnn_trainer/engine.py]: ok then instead all stuff I did basically I have to add the following line in fit "history = {"train":train_metrics, "val" : valid_metrics}" right

File: pytorch_cnn_trainer/engine.py [hunk 5]
```diff
@@ -333,9 +333,19 @@ def fit(
...
 
+        loss , top1_acc , top5_acc = [train_metrics[i] for i in ('loss','top1','top5')]
+        history = {"train" : {"top1_acc" : top1_acc , "top5_acc" : top5_acc, "loss" : loss},"val" : { }}
+
+
         print()
         print(f"Validating Epoch = {epoch}")
         valid_metrics = val_step(model, valid_loader, criterion, device, num_batches, log_interval)
+        
+        loss_v , top1_acc_v , top5_acc_v = [train_metrics[i] for i in ('loss','top1','top5')]
```
Comment (premalrupnur) [inline pytorch_cnn_trainer/engine.py]: ohh sorry I haven't changed train_metrics here , I just did copy paste so forgot

Comment (oke-aditya) [reply pytorch_cnn_trainer/engine.py]: Looks better now, this is cleaner right?

Can you add code to plot this using `matplotlib`. This can go in seperate file called `plotters.py` (name suggestions needed !)

File: tests/test_fit _metrcis.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                valid_loader,
+                criterion,
+                device,
+                optimizer,
+                num_batches=10,
+                grad_penalty=True,
+                use_fp16=True,
+            )
+
+            # Testing keys of history dictionary
```
Comment (oke-aditya) [inline tests/test_fit _metrcis.py]: This looks like code duplication below.
I guess simpler will be to use either history object using `if` `else` and run the same tests.

This will reduce the code.

File: tests/test_fit _metrcis.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        t1,t5,l = list(history2[str1].keys())
+        assert(('top1_acc','top5_acc','loss')==(t1,t5,l))
+        assert((epoch,epoch,epoch)==(len(history2[str1][t1]),len(history2[str1][t5]),len(history2[str1][l])))
+            
+        # Testing the metrics for val set
+        t1,t5,l = list(history2[str2].keys())
+        assert(('top1_acc','top5_acc','loss')==(t1,t5,l))
+        assert((epoch,epoch,epoch)==(len(history2[str2][t1]),len(history2[str2][t5]),len(history2[str2][l])))
+
+
```
Comment (oke-aditya) [inline tests/test_fit _metrcis.py]: Can you keep only 1 blank line if needed to seperate.

File: pytorch_cnn_trainer/engine.py [hunk 8]
```diff
@@ -348,7 +392,11 @@ def fit(
 
         print("Done Training, Model Saved to Disk")
 
-    return True  # For now, we should probably return an history object like keras.
+    history = {
```
Comment (oke-aditya) [inline pytorch_cnn_trainer/engine.py:395]: Looks Good

File: pytorch_cnn_trainer/engine.py [hunk 9]
```diff
@@ -419,11 +467,15 @@ def train_sanity_fit(
...
+                    f"Training Sanity check passed in time {train_sanity_end - train_sanity_start} !!"
+                )
                 return True
 
     train_sanity_end = time.time()
-    print(f"Training Sanity check passed in time {train_sanity_end - train_sanity_start} !!")
+    print(
+        f"Training Sanity check passed in time {train_sanity_end - train_sanity_start} !!"
+    )
     return True
```
Comment (oke-aditya) [inline pytorch_cnn_trainer/engine.py:479]: Can the same logic be applied to `sanity_fit` too ?

Since users will expect the same API for both.

File: tests/test_fit _metrcis.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+from tqdm import tqdm
+import torchvision.transforms as T
+from pytorch_cnn_trainer import dataset
+from pytorch_cnn_trainer import model_factory
+from pytorch_cnn_trainer import utils
+from pytorch_cnn_trainer import engine
+from torch.optim.swa_utils import SWALR
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
```
Comment (oke-aditya) [inline tests/test_fit _metrcis.py]: nit: reame the file from `test_fit _metrics.py` to `test_fit_metrics.py`

**Review by oke-aditya — CHANGES_REQUESTED** [review]
75 % done. You are almost there. I need little bit of changes.

File: pytorch_cnn_trainer/engine.py (hunk 9 above)
Comment (premalrupnur) [reply pytorch_cnn_trainer/engine.py:479]: yes I would implement it

**Review by oke-aditya — APPROVED** [review]

**Review by oke-aditya — APPROVED** [review]
I need to test them though

**Comment by oke-aditya** [comment]
Thanks a lot for the PR @premalrupnur 
I haven't tested this yet.
But I guess I will merge this since it looks good.
I will test this before next PyPi release

File: pytorch_cnn_trainer/engine.py [hunk 11]
```diff
@@ -358,6 +392,10 @@ def fit(
 
         print("Done Training, Model Saved to Disk")
 
+    history = {
+        "train": {"top1_acc": top1_acc, "top5_acc": top5_acc, "loss": loss},
+        "val": {"top1_acc": loss_v, "top5_acc": top5_acc_v, "loss": top1_acc_v},
```
Comment (premalrupnur) [inline pytorch_cnn_trainer/engine.py:397]: Here I have shuffled the value  for top1_acc_v and loss_v would commit this change in next pr

Comment (oke-aditya) [reply pytorch_cnn_trainer/engine.py:397]: Yeah that's a bad miss from our side, we should fix this up 😄
